### ISSUE TITLE
Add support for --build-args

### ISSUE THREAD

**Issue by kismet010** [docker/compose#2111]
As this docker/docker#15182 functionality has been merged, it would be great to manage with Compose

I.e.:

Dockerfile with default variables
```
ARG vhost=default.com
RUN echo $vhost > ...
```

Overriding through docker-composer.yml
```
web:
  build: ...
  arguments:
    - vhost: dev.com
```

Or an hierarchy of extended environments
```
production:
  extends: web
  arguments:
    - vhost: my-final-domain.com
```

**Comment by dnephin** [comment]
I've been wanting to make a change to the build config, something like this:

```yaml

web:
  build:
    context: .
    dockerfile: dockerfiles/path/Dockerfile
    arguments:
      arg1: ...
```

**Comment by dnephin** [comment]
I created a proposal for this configuration change at #2163

**Comment by smboykin** [comment]
+1 on this feature request

**Comment by snelis** [comment]
+1 on passing environment variables onto the build step via compose

**Comment by lsborg** [comment]
+1

**Comment by Silex** [comment]
+1

**Comment by marcellodesales** [comment]
I'm stuck at Compose right now... :( Please please add this support...

I can sucessfully build with the the --build-arg because I need the HTTP_PROXY on my build servers... :(

```sh
[root@pppdc9prda2x npmo-docker-compose]# docker build --build-arg HTTP_PROXY=$HTTP_PROXY -t configure -f ./configure.Dockerfile .
Sending build context to Docker daemon 141.8 kB
Step 1 : FROM node:4-onbuild
# Executing 3 build triggers...
Step 1 : COPY package.json /usr/src/app/
 ---> Using cache
Step 1 : RUN npm install
 ---> Using cache
Step 1 : COPY . /usr/src/app
 ---> 52f129b79095
Removing intermediate container 00cb3b4651d4
Step 2 : CMD npm run configure
 ---> Running in 7d0642695c58
 ---> db140f15fe39
Removing intermediate container 7d0642695c58
Successfully built db140f15fe39
```

Build from compose
----

I still get a build stuck trying to connect to resources... I would propose:

```
  build_args:
    - HTTP_PROXY=$HTTP_PROXY
```

```yml
licenseData:
  image: busybox
  volumes:
    - .license.json:/usr/src/app/.license.json

configure:
  build: .
  dockerfile: configure.Dockerfile
  build_args:
    - HTTP_PROXY=$HTTP_PROXY
  volumes_from:
    - "licenseData"
  environment:
    - HTTP_PROXY=$HTTP_PROXY
```

The build just hangs...

```sh
[root@pppdc9prda2x npmo-docker-compose]# docker-compose -f configure.yml up
Starting npmodockercompose_licenseData_1
Building configure
Step 1 : FROM node:4-onbuild
# Executing 3 build triggers...
Step 1 : COPY package.json /usr/src/app/
 ---> Using cache
Step 1 : RUN npm install
 ---> Running in ba76ce5f09f6
npm info it worked if it ends with ok
npm info using npm@2.14.7
npm info using node@v4.2.2
npm WARN package.json npmo-docker@1.0.0 No repository field.
npm WARN package.json npmo-docker@1.0.0 No README data
npm WARN package.json npmo-docker@1.0.0 No bin file found at ./bin/npme-docker-compose.js
npm info preinstall npmo-docker@1.0.0
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/chalk
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/inquirer
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/lodash
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/mustache
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/request
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/yargs
npm info attempt registry request try #1 at 9:59:27 PM
npm http request GET https://registry.npmjs.org/standard
```

**Comment by trcarden** [comment]
+1

**Comment by mixja** [comment]
+1

**Comment by wr0ngway** [comment]
+1

**Comment by McSlow** [comment]
+1

**Comment by wricardo** [comment]
+1

**Comment by munhitsu** [comment]
+1

**Comment by jgeskens** [comment]
+1

**Comment by TeChn4K** [comment]
Please ... +1 !

**Comment by slashrsm** [comment]
+1 :)

**Comment by k2tzumi** [comment]
+1

**Comment by jrosadohp** [comment]
+1 !!

**Comment by funkytaco** [comment]
:+1:

**Comment by levivm** [comment]
+1

**Comment by Jason-Grant** [comment]
+1

**Comment by dirkmoors** [comment]
+1

**Comment by B-Galati** [comment]
+1 !

**Comment by raphink** [comment]
+1

**Comment by saulshanabrook** [comment]
FYI for everyone +1 ing, this is currently being implemented in #2540, which is blocked waiting on #2421 to be merged.

**Comment by bikong0411** [comment]
+1

**Comment by wil93** [comment]
Can someone lock this conversation to members only? :sweat: @dnephin?

**Comment by gmjonker** [comment]
+1

**Comment by marcellodesales** [comment]
Awesome!

**Comment by dnephin** [comment]
Closed by #2540

**Comment by ericm24** [comment]
+1

**Comment by aanand** [comment]
+1

**Comment by aanand** [comment]
1.6.0 RC1 is now out with support for build args:

https://github.com/docker/compose/releases/tag/1.6.0-rc1

**Comment by ericm24** [comment]
@aanand: Thanks for the change!  

I pulled down the change and modified my docker-compose.yml to what I think should be correct version: 2 format, and am getting the following when I run:

docker-compose ps

ERROR: client is newer than server (client API version: 1.22, server API version: 1.21)

Is there a a 1.22 RC .ISO image?

Eric

**Comment by dnephin** [comment]
Assuming you mean a boot2docker iso, there is one for the 1.10rc1 release https://github.com/tianon/boot2docker-legacy/releases/tag/v1.10.0-rc1

**Comment by mrlamroger** [comment]
Would love 1.6.0rc1 to be pushed to pypi ;) Congrats btw!

**Comment by dnephin** [comment]
It's on pypi, it's just hidden (https://pypi.python.org/pypi/docker-compose/1.6.0rc1) because the setup.py is wrong. It depends on docker-py 1.7.0, but we released before it was available. If you install it from pypi make sure to grab docker-py 1.7.0-rc1

**Comment by sikhote** [comment]
After downloading bootdocker.iso from @dnephin 's link, I used this command to start the default docker VM:

`docker-machine create -d virtualbox --virtualbox-boot2docker-url=file://$HOME/Downloads/boot2docker.iso default`

This assumes the download is in your user/Downloads folder.

**Comment by sikhote** [comment]
1.6 RC1 works great so far, thanks!

Is there any support for adding build args to the docker-compose build command? Such as...

`docker-compose -f docker/docker-compose.yml --build-arg NODE_ENV=production build`

**Comment by ericm24** [comment]
+1 on sikhote above.   I would have thought that:

docker-compose -f docker/docker-compose.yml --build-arg NODE_ENV=production build

would be included....the --build-arg 'ness would be added to docker-compose for BOTH:
-build
-run

**Comment by saulshanabrook** [comment]
Adding `--build-arg` would also be good for me, but it might be problematic because your are often building multiple images. Would the `--build-arg` be applied to all images that are being built? That seems a bit heavy handed, but I also would find it useful.

**Comment by sikhote** [comment]
Yes, I was imagining across all images as well. For the moment, the solution is to extend or overwrite a base docker-compose.yml, but it's nice to avoid those extra files for (in my case) a single build argument.

**Comment by ericm24** [comment]
My situation.  I have a jenkins dockerfile that includes an installation of postfix.  I need postfix so jenkins has an smtp mail to send failed builds.    The Postfix config needs the HOST-IP set in it to work correctly....   I do not (WILL-NOT) want to hard-code the host-ip address in my docker-compose.yml, Dockerfile, etc....

Given the docker-compose.yml snippet below:

version: 2
services:
  jenkins:
    build:
      context: ./jenkins
      dockerfile: Dockerfile
      args:
        - DOCKER_HOST_IP=${DOCKER_HOST_IP}

    ports:
      - "18080:8080"

    environment:
      - DOCKER_HOST_IP=${DOCKER_HOST_IP}

    volumes:
      - /mnt/dockervolumes/data/jenkins:/var/jenkins_home



>> I would expect the ARGS to be applied for the BUILD, so that they are available for RUN. <<

IMHO:  Being able to over-ride --build-args on a docker-compose build ...  command adds more power & flexibility to it, and I'd expect that to override whatever was in a docker-compose.yml....

Right now, with my current configuration, I cannot use docker-compose for ALL of my apps in my docker-compose.yml. (jenkins, nexus, github, sonar, selenium).   I have run docker-compose up to get all of them up initially....  However, given the missing --build-args, (Jenkins will be wrong), so I then have to do a subsequent down/build (4 Jenkins)/up to get them (rest + jenkins) correct.

docker-compose up       // Gets all up, but jenkins will be wrong....
docker-compose down   // Shuts down ALL apps down,  I just started.
docker rm `docker ps -aq -f  status=exited`  // Removes containers I just created... [ jenkins is wrong, others are OK... ].
docker build --build-arg DOCKER_HOST_IP=$DOCKER_HOST_IP --rm=true --no-cache --tag=dockercitoolstackmaster_jenkins .  | tee /data/txt/dockerBuildLog.txt  // Rebuilds Jenkins Image w/build-args

docker-compose up  // All the kings horses & all the kings men are now right once-again....Now we've got ALL apps built correctly w/Jenkins using --build-arg.







docker build --build-arg DOCKER_HOST_IP=$DOCKER_HOST_IP --rm=true --no-cache --tag=dockercitoolstackmaster_jenkins .  | tee /data/txt/dockerBuildLog.txt   // rebuilds my jenkins image with the --build-arg

**Comment by marcellodesales** [comment]
@ericm24 Would you mind sharing a GitHub Repo with your `docker-compose.yml` file with what you have described? If you got this right, It will definitely help a lot of people (including myself).

**Comment by ericm24** [comment]
@marcellodesales Yes, I will share a repo.   Will let you know when I get it pushed up.

**Comment by dnephin** [comment]
`docker-compose build --build-arg`  would not work (for the same reason that `docker-compose up --some-run-flag` doesn't work. The values are applied to **every** service. If you don't have the `ARG` defined in every dockerfile, it's an error. Having an identical `ARG` in every service seems like a very uncommon use case.

This is just how Compose is defined, any differences between services need to be in the Compose file (not applied through command line flags).  Using variables is a good option.  As I called out in #1455 Compose is primarily a tool for running services, not building images. It's not going to be possible to support every unique build scenario because the tool is not designed for that purpose. We can only really support the most common build scenarios.

@ericm24 to handle your case, why not just run:

```
DOCKER_HOST_IP=$DOCKER_HOST_IP docker-compose build
```

That should work

**Comment by marcellodesales** [comment]
I agree with @dnephin... I use `docker-compose` as a means of storing the entire runtime setting, without initially depending on the environment. I would think the external environments definitely come handy for overrides...

**Comment by ericm24** [comment]
Thanks for the help!  I've made some progress on this, sorry it took a while.  A couple of items:

#1  I'm running into the ERROR: client is newer than server (client API version: 1.22, server API version: 1.21).  However, this is on my Ubuntu Boxes which DO NOT use Virtualbox VMs.   I have a copy of the correct .iso file, but how do I make changes to use the correct .iso file on Docker on Ubuntu?  Please pardon my newbie'ness here.....

#2  @marcellodesales:  Please see git repo: https://github.com/ericm24/repo1.git
for the example files for this.   However, I've run into a problem for the nexus build.  The docker-repo for centos7 appears unavailable....  So I have not been able to test this fully.... I ~think~ it will work...

This network-error started happening last on or about Weds [ 27-Jan ].  Was working fine before that, AFAIK I did not change anything on my end for network proxy, etc....weird.  Error Text is:
Building nexus
Step 1 : FROM centos:7
Pulling repository docker.io/library/centos
ERROR: Service 'nexus' failed to build: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Network timed out while trying to connect to https://index.docker.io/v1/repositories/library/centos/images. You may want to check your internet connection or if you are behind a proxy.

#3 @dnephin using:
DOCKER_HOST_IP=$DOCKER_HOST_IP docker-compose build

Agreed!  I think would be OK.  I do not think docker-compose should have to juggle command-line args with what is in a docker-compose.yml.   (ie what if they don't match?)

For me, I do not want to have to issue a special/separate 'docker-compose build..'  or 'docker build...' for 1 of my apps in my docker-compose group...  

What I was hoping would work would be to execute a:

docker-compose build jenkins

and docker-compose would ~handle~ the ARG portion for the jenkins service found in the docker-compose.yml

or in the case of the first-time build using:

docker-compose up

Would be all that is needed....

Again, docker-compose would ~handle~ service/application-level ARG items found in the docker-compose.yml

Basically docker-compose would have to make a ~pass~ at the application/service level for any "args" items and then add them to the respective build command for that specific application/service.

When you make the statement:
"...compose is primarily a tool for running services, not building images..."   

compose is invoking a build on an image in some cases (ie docker-compose up for the first time)... or if the image file has been deleted...

**Comment by ikabar-tracxpoint** [comment]
I would like to see this feature. don't agree with @dnephin many having an easy way to build your multi container application bring great value.

**Comment by nazneen84** [comment]
+2 :)

**Comment by mixja** [comment]
@ericm24 - I'm not sure why you would want something as ephemeral as an IP address in your Docker image build process.  

This seems like a runtime concern, which you either deal with externally by injecting into the environment (as per @dnephin), or you could do some introspection to determine this in an entry point script that starts up your container.

**Comment by m1rza-s** [comment]
Has this been implemented?
